File: dev/src/samples/NetSamplesVB/Samples/Profiles/ProfileCreationCommands.vb
Imports System

Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.EditorInput
Imports Autodesk.AutoCAD.Geometry
Imports Autodesk.AutoCAD.Runtime
Imports Autodesk.Civil
Imports Autodesk.Civil.DatabaseServices

<Assembly: CommandClass(
GetType(Autodesk.CivilizedDevelopment.ProfileCreationCommands))> 

Namespace Autodesk.CivilizedDevelopment
  Public Class ProfileCreationCommands
    Inherits SimpleDrawingCommand
    <CommandMethod("CDS_CreateMultipleProfileViews")> _
    Public Sub CDS_CreateMultipleProfileViews()
      Using tr As Transaction = startTransaction()
        Dim created As Boolean = createMultipleProfiles()
        If created Then
          tr.Commit()
        End If
      End Using
    End Sub



    Private Function createMultipleProfiles() As Boolean
      If noAlignmentSelected() Then
        Return False
      End If
      If noInsertionPointSelected() Then
        Return False
      End If
      createProfileViews()
      Return True
    End Function



    Private Function noAlignmentSelected() As Boolean
      _alignmentId = selectEntity(Of Alignment)(
        vbLf & "Select alignment:")
      Return _alignmentId.IsNull
    End Function



    Private Function noInsertionPointSelected() As Boolean
      Dim selected As Boolean = False
      Dim result As PromptPointResult = _editor.GetPoint(
        vbLf & "Select insertion point")
      If result.Status = PromptStatus.OK Then
        _insertionPoint = result.Value
        selected = True
      End If
      Return Not selected
    End Function



    Private Sub createProfileViews()
      ProfileView.CreateMultiple(_alignmentId, _insertionPoint,
                                 _creationOptions)
    End Sub



    Private ReadOnly Property _creationOptions() _
        As MultipleProfileViewsCreationOptions
      Get
        Dim options As New MultipleProfileViewsCreationOptions()
        options.DrawOrder = ProfileViewPlotType.ByRows
        options.GapBetweenViewsInColumn = 1000
        options.GapBetweenViewsInRow = 1000
        options.MaxViewInRowOrColumn = 3
        options.StartCorner = ProfileViewStartCornerType.UpperLeft
        Return options
      End Get
    End Property

    Private Function selectEntity(Of T)(prompt As String) As ObjectId
      Dim options As New PromptEntityOptions(prompt)
      options.SetRejectMessage(vbLf & "Incorrect entity type.")
      options.AddAllowedClass(GetType(T), True)
      Dim result As PromptEntityResult = _editor.GetEntity(options)
      If result.Status = PromptStatus.OK Then
        Return result.ObjectId
      End If
      Return ObjectId.Null
    End Function




    Private _alignmentId As ObjectId
    Private _insertionPoint As Point3d
  End Class
End Namespace

File: dev/src/samples/NetSamples/Samples/Profiles/ProfileCreationCommands.cs
﻿using System;

using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.EditorInput;
using Autodesk.AutoCAD.Geometry;
using Autodesk.AutoCAD.Runtime;
using Autodesk.Civil;
using Autodesk.Civil.DatabaseServices;

[assembly: CommandClass(
    typeof(Autodesk.CivilizedDevelopment.ProfileCreationCommands))]

namespace Autodesk.CivilizedDevelopment
{
    public class ProfileCreationCommands : SimpleDrawingCommand
    {
        [CommandMethod("CDS_CreateStackedProfileViews")]
        public void CDS_CreateStackedProfileViews()
        {
            using (Transaction tr = startTransaction())
            {
                bool created = createStackedProfiles();
                if (created)
                {
                    tr.Commit();
                }
            }
        }

        [CommandMethod("CDS_CreateMultipleProfileViews")]
        public void CDS_CreateMultipleProfileViews()
        {
            using (Transaction tr = startTransaction())
            {
                bool created = createMultipleProfiles();
                if (created)
                {
                    tr.Commit();
                }
            }
        }

        private bool createStackedProfiles()
        {
            if (noAlignmentSelected()) return false;
            if (noInsertionPointSelected()) return false;
            createStackedProfileViews();
            return true;
        }

        private bool createMultipleProfiles()
        {
            if (noAlignmentSelected()) return false;
            if (noInsertionPointSelected()) return false;
            createProfileViews();
            return true;
        }

        private bool noAlignmentSelected()
        {
            _alignmentId = selectEntity<Alignment>("\nSelect alignment:");
            return _alignmentId.IsNull;
        }        

        private bool noInsertionPointSelected()
        {
            bool selected = false;
            PromptPointResult result = 
                _editor.GetPoint("\nSelect insertion point");
            if (result.Status == PromptStatus.OK)
            {
                _insertionPoint = result.Value;
                selected = true;
            }
            return !selected;
        }

        private void createStackedProfileViews()
        {
            ProfileView.CreateMultiple(_alignmentId, _insertionPoint, 
                _stackedOptions, _creationOptions);
        }

        private void createProfileViews()
        {
            ProfileView.CreateMultiple(_alignmentId, _insertionPoint,
                _creationOptions);
        }

        private MultipleProfileViewsCreationOptions _creationOptions
        {
            get
            {
                MultipleProfileViewsCreationOptions options =
                    new MultipleProfileViewsCreationOptions();
                options.DrawOrder = ProfileViewPlotType.ByRows;
                options.GapBetweenViewsInColumn = 1000;
                options.GapBetweenViewsInRow = 1000;
                options.MaxViewInRowOrColumn = 3;
                options.StartCorner = 
                    ProfileViewStartCornerType.UpperLeft;
                return options;
            }
        }

        private StackedProfileViewsCreationOptions _stackedOptions
        {
            get
            {
                StackedProfileViewsCreationOptions options =
                    new StackedProfileViewsCreationOptions(_defaultPVId,
                        _defaultPVId, _defaultPVId);
                return options;
            }
        }

        private ObjectId _defaultPVId
        {
            get
            {
                return _civildoc.Styles.ProfileViewStyles[0];
            }
        }

        private ObjectId selectEntity<T>(string prompt)
        {
            PromptEntityOptions options =
                new PromptEntityOptions(prompt);
            options.SetRejectMessage("\nIncorrect entity type.");
            options.AddAllowedClass(typeof(T), true);
            PromptEntityResult result = _editor.GetEntity(options);
            if (result.Status == PromptStatus.OK)
            {
                return result.ObjectId;
            }
            return ObjectId.Null;
        }


        ObjectId _alignmentId;
        Point3d _insertionPoint;
    }
}